### PR TITLE
#1340 | Fixing the fix_table method creating unnecessary tbody elemen…

### PR DESCRIPTION
**Pull request by simonberger**
…ts around every single tr

### CODE REVIEW

**Comment by simonberger** [comment]
One side effect I noted about that fix.
If you have a buggy input like
```
<table>
  <td>bad placed tag 1</td>
  <p>bad placed tag 2</p>
  <tr>
    <td>row 1 column 1</td>
  </tr>
</table>
```

tbody will be placed also around the `td` and `p` tags.
I wrote about this [here](https://github.com/dompdf/dompdf/issues/1363#issuecomment-275194149)

I don't think that has any worse effect to the old behaviour. Dompdf can't handle that bad input very well and maybe shouldn't anyway. But I didn't test this much.

**Comment by bsweeney** [comment]
Do you think that instead of selecting all children we should select all direct descendent TR elements? It seems unlikely, but what if somebody tries to render a document that contains a table similar to the following:

```
<table>
  <thead>
    ...
  </thead>
  <tr>...</tr>
    ..
</table>
```

It seems unlikely, except that we see all kinds of wacky HTML.

**Comment by simonberger** [comment]
You are correct. That's a case not being handled well. Also some TRs included in a tbody element following some plain rows wouldn't work.
These cases should be at least valid html.

First I wanted to just skip those leading elements resulting in a more complex method but then got a better solution. We already start with the leading tr dom-position having those possible previous siblings.
So just continue from this point walking through the siblings and still append the tbody to the table at the end.

**Comment by simonberger** [comment]
Sorry, my commits sometimes follow a learning process ;)
Thought again and realized the situation addressed in the comment before still isn't fixed if the leading tr followed by a tbody.
```
<table>
		<tr><td>1</td></tr>
		<tbody><tr><td>2</td></tr></tbody>
		<p>missplaced tag</p>
		<tr><td>3</td></tr>
</table>
```

Results now in

```
<table>
		<tbody><tr><td>1</td></tr></tbody>
		<tbody><tr><td>2</td></tr></tbody>
		<p>missplaced tag</p>
		<tbody><tr><td>3</td></tr></tbody>
</table>
```

**Comment by bsweeney** [comment]
I think we're ok to ignore obviously wrong table formatting such as a P as a child of the TABLE element. If someone's HTML is going to be that poorly formed they should use the HTML5 parser. Since a TR is appropriately a child of the table element, but for the purposes of parsing our document we're moving inside a TBODY, I think this is the only exception we want to handle.

Whenever we get around to fully supporting anonymous table objects we can deal with exceptions at that point.

**Comment by simonberger** [comment]
Ignoring the P element was just meant as a "bonus" of the last commit.
The idea was instead to support the (repeated) mix of TR and TBODY direct child elements of a table (the tbody before that missplaced p in the example) beside your correct argument a THEAD followed by TRs would produce wrong restults.

You are free to skip the last two commits but I think the code additions and performance losses are too tiny to not support those unlikely but valid html case.

**Comment by trevorclarke** [comment]
Hi,Not sure if it helps the discussion, at present I regularly use domPDF to process html tables that use thead and tbody. I've not ever used the tfooter tag.tablethead    tr        th/th    /tr/theadtbody    tr        td/td    /tr/tbody/tableThese can and do sometimes have associated css with them.Regards,Trevor Clarke BSc (hons), MCMI Director0780 90 623 68Chittak Ltd Office: 01202 611 711  Skype: chittakwebdevelopment  Website: www.chittak.co.ukRaR Recommended


-------- Original Message --------
Subject: Re:_[dompdf/dompdf]_#1340_|_Fixing_the_fix_table
_method_creating_unnecessary_tbody_elemen…_(#1357)

File: src/Frame/FrameTree.php [hunk 1]
```diff
@@ -160,17 +160,35 @@ protected function fix_tables()
...
             $table->parentNode->insertBefore($caption, $table);
         }
 
-        $rows = $xp->query("//table/tr");
-        foreach ($rows as $row) {
-            $tbody = $this->_dom->createElement("tbody");
-            $tbody = $row->parentNode->insertBefore($tbody, $row);
-            $tbody->appendChild($row);
+        $firstRows = $xp->query('//table/tr[1]');
+        $tbodyInsertPosition = null;
```
Comment (bsweeney) [inline src/Frame/FrameTree.php]: Leftover variable?

Comment (simonberger) [inline src/Frame/FrameTree.php]: Yay correct, missed this from a previous way.

**Pull request merged**